main.go

```diff
@@ -504,20 +524,40 @@ func CreateBackup(backupName string, etcdCACert, etcdCert, etcdKey, endpoints st
 			}).Warn("changing permission of the compressed snapshot failed")
 			continue
 		}
+		break
+	}
+	return
+}
 
-		if server.Backup {
-			// If folder is specified, prefix the file with the folder
-			if len(server.Folder) != 0 {
-				compressedFile = fmt.Sprintf("%s/%s", server.Folder, compressedFile)
-			}
-			err = uploadBackupFile(svc, server.BucketName, compressedFile, compressedFilePath)
-			if err == nil {
-				return nil
-			}
+func CreateS3Backup(backupName, compressedFilePath string, bc *backupConfig) error {
+	// If the minio client doesn't work now, it won't after retrying
+	client, err := minioClientFromConfig(bc)
+	if err != nil {
+		return err
+	}
+	compressedFile := filepath.Base(compressedFilePath)
+	// If folder is specified, prefix the file with the folder
+	if len(bc.Folder) != 0 {
+		compressedFile = fmt.Sprintf("%s/%s", bc.Folder, compressedFile)
+	}
+	// check if it exists already in the bucket, and if versioning is disabled on the bucket. If an error is detected,
+	// assume we aren't privy to that information and do multiple uploads anyway.
+	info, _ := client.StatObject(context.TODO(), bc.BucketName, compressedFile, minio.StatObjectOptions{})
+	if info.Size != 0 {
+		versioning, _ := client.GetBucketVersioning(context.TODO(), bc.BucketName)
+		if !versioning.Enabled() {
+			log.WithFields(log.Fields{
+				"name": backupName,
+			}).Info("Skipping upload to s3 because snapshot already exists and versioning is not enabled for the bucket")
+			return nil
 		}
-		break
 	}
-	return err
+
+	err = uploadBackupFile(client, bc.BucketName, compressedFile, compressedFilePath, s3Retries)
+	if err == nil {
```

Comment:
You're never returning `err` here if err != nil (if uploading the backup file fails). Should be fixed

Reply:
Good catch